spec.html

```diff
@@ -7573,134 +7573,44 @@ <h1>GetGlobalObject ( )</h1>
+      <p>Host environments are not required to treat Jobs uniformly with respect to scheduling. For example, web browsers treat Promise-handling Jobs as microtasks. Future features may add Jobs that are treated as tasks.</p>
     </emu-note>
-    <p>The following abstract operations are used to create and manage Jobs and Job Queues:</p>
+
+    <emu-clause id="sec-hostenqueuepromisejob" aoid="HostEnqueuePromiseJob">
+      <h1>HostEnqueuePromiseJob ( _job_ )</h1>
+
+      <p>HostEnqueuePromiseJob is a host-defined abstract operation that schedules the Job abstract closure _job_ to be performed, at some future time. The abstract closures used with this algorithm are intended to be related to the handling of Promises, or otherwise, to be scheduled with equal priority to Promise handling operations.</p>
+
+      <p>The implementation of HostEnqueuePromiseJob must conform to the requirements in <emu-xref href="#sec-jobs"></emu-xref>. Additionally, Jobs must be scheduled in FIFO order, with Jobs running in the same order as the HostEnqueuePromiseJob invocations which scheduled them.</p>
+    </emu-clause>
 
     <emu-clause id="sec-enqueuejob" aoid="EnqueueJob">
-      <h1>EnqueueJob ( _queueName_, _job_, _arguments_ )</h1>
-      <p>The EnqueueJob abstract operation requires three arguments: _queueName_, _job_, and _arguments_. It performs the following steps:</p>
+      <h1>EnqueueJob ( _kind_, _job_ )</h1>
+      <p>The abstract operation EnqueueJob takes two arguments, _kind_ and an abstract closure _job_, and performs the following steps:</p>
       <emu-alg>
-        1. Assert: Type(_queueName_) is String and its value is the name of a Job Queue recognized by this implementation.
-        1. Assert: _job_ is the name of a Job.
-        1. Assert: _arguments_ is a List that has the same number of elements as the number of parameters required by _job_.
-        1. Let _callerContext_ be the running execution context.
-        1. Let _callerRealm_ be _callerContext_'s Realm.
-        1. Let _callerScriptOrModule_ be _callerContext_'s ScriptOrModule.
-        1. Let _pending_ be PendingJob { [[Job]]: _job_, [[Arguments]]: _arguments_, [[Realm]]: _callerRealm_, [[ScriptOrModule]]: _callerScriptOrModule_, [[HostDefined]]: *undefined* }.
-        1. Perform any implementation or host environment defined processing of _pending_. This may include modifying the [[HostDefined]] field or any other field of _pending_.
-        1. Add _pending_ at the back of the Job Queue named by _queueName_.
+        1. Assert: _job_ is a Job abstract closure.
+        1. If _kind_ is ~promise~, then perform ! HostEnqueuePromiseJob(_job_).
```

Comment:
Can we just assert on `_kind_` here for now? Possibly with a `Note` explaining what's going on.

Reply:
I still weakly prefer keeping it a conditional. I guess the change will be pretty local to merge WeakRefs, but I don't really see a compelling reason to change it to an assert.

Reply:
I just don't like non-exhaustive switches, really.